### PR TITLE
PTypes 31: ReadRows uses PType

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

File: hail/src/main/scala/is/hail/HailContext.scala [hunk 1]
```diff
@@ -614,12 +615,12 @@ class HailContext private(val sc: SparkContext,
...
   def readRows(
     path: String,
-    t: TStruct,
+    t: PStruct,
     codecSpec: CodecSpec,
     partFiles: Array[String],
     requestedType: TStruct
   ): ContextRDD[RVDContext, RegionValue] = {
-    val makeDec = codecSpec.buildDecoder(t.physicalType, requestedType.physicalType)
+    val makeDec = codecSpec.buildDecoder(t, requestedType)
```
Comment (cseed) [inline hail/src/main/scala/is/hail/HailContext.scala]: Great.  And eventually `t` will become an encoded type and the decoder will specify its physical type.

File: hail/src/main/scala/is/hail/expr/types/MatrixType.scala [hunk 2]
```diff
@@ -17,6 +18,12 @@ class MatrixTypeSerializer extends CustomSerializer[MatrixType](format => (
 object MatrixType {
   val entriesIdentifier = "the entries! [877f12a8827e18f61222c6c8c5fb04a8]"
 
+  def getRowType(rvRowType: PStruct): PStruct = rvRowType.selectFields(Set(entriesIdentifier), keep = false)
+  def getEntryArrayType(rvRowType: PStruct): PArray = rvRowType.field(entriesIdentifier).typ.asInstanceOf[PArray]
+  def getEntriesType(rvRowType: PStruct): PStruct = rvRowType.selectFields(Set(entriesIdentifier), keep = true)
```
Comment (cseed) [inline hail/src/main/scala/is/hail/expr/types/MatrixType.scala]: `getEntriesType` seems weird.  It's the row with just the single field entries?  I assume that's for read/write split?  I don't think it should be in MatrixType.

Comment (cseed) [reply hail/src/main/scala/is/hail/expr/types/MatrixType.scala]: Or it needs a clearer name, like `getEntriesOnlyRowType`.

File: hail/src/main/scala/is/hail/io/RowStore.scala [hunk 3]
```diff
@@ -1665,15 +1665,15 @@ class RichContextRDDRegionValue(val crdd: ContextRDD[RVDContext, RegionValue]) e
     val nPartitions = crdd.getNumPartitions
     val d = digitsNeeded(nPartitions)
 
-    val fullRowType = t.rvRowType
-    val fullRowPType = fullRowType.physicalType
-    val rowsRVType = t.rowType
-    val localEntriesIndex = t.entriesIdx
-    val entriesRVType = t.entriesRVType
+    val fullRowType = t.rowType
```
Comment (cseed) [inline hail/src/main/scala/is/hail/io/RowStore.scala]: This looks wrong.  I thought "full" meant including entries, but t.rowType doesn't include entries.  Maybe you just need to rename fullRowType to rowType?

File: hail/src/main/scala/is/hail/HailContext.scala (hunk 1 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/HailContext.scala]: yeah, exactly.

Comment (tpoterba) [reply hail/src/main/scala/is/hail/HailContext.scala]: this stuff is one of the corners where the type/ptype system breaks down currently. we need ETypes.

File: hail/src/main/scala/is/hail/io/RowStore.scala (hunk 3 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/io/RowStore.scala]: t is the RVDType, so the rowType does include entries

Comment (tpoterba) [reply hail/src/main/scala/is/hail/io/RowStore.scala]: then I use the MatrixType operations to get out the entries-only and no-entries types

**Comment by tpoterba** [comment]
I also realized I need to add the Legacy reader to make this stuff work - I've changed the json serialization of the RVDSpec

**Comment by tpoterba** [comment]
OK, I'm going to close this and push this functionality later. First I want to support the legacy stuff.

**Pull request closed without merging**